### PR TITLE
Implementation of the Temporal Dictionary Ensemble (TDE) classifier

### PR DESCRIPTION
**Pull request by MatthewMiddlehurst**
#### What does this implement/fix? Explain your changes.

Implements the Temporal Dictionary Ensemble (TDE) dictionary based time series classifier. Related paper accepted to ECML PKDD 2020 and currently in the process of being published.

Also includes some fixes, tests and documentation for BOSS. I have added the dictionary based transforms to the estimator overview files, if these are not in the correct location let me know. 

#### PR checklist for new estimators 

- \[x] I have added unit tests and made sure they pass locally. 
 
- \[x] I have updated the existing example notebooks or provided a new one to showcase how my algorithm works.
   
- \[x] I have updated sktime's 
[estimator overview](https://github.com/alan-turing-institute/sktime/blob/master/ESTIMATOR_OVERVIEW.md) and I'm
 committed to maintain my contribution and provide bug fixes if necessary.

### CODE REVIEW

**Comment by mloning** [comment]
@MatthewMiddlehurst Thanks for the new stuff! Let me know when you're ready and I can review it!

**Comment by MatthewMiddlehurst** [comment]
Ready to go from my end. If there are any changes you need let me know.

File: sktime/classification/dictionary_based/_tde.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+                preds = gp.predict(possible_parameters)
+                parameters = possible_parameters.pop(rng.choice(
+                    np.flatnonzero(preds == preds.max())))
+
+            subsample = rng.choice(self.n_instances, size=subsample_size,
+                                   replace=False)
+            X_subsample = X.iloc[subsample, :]
+            y_subsample = y[subsample]
+
+            tde = IndividualTDE(parameters[0], parameters[1],
```
Comment (mloning) [inline sktime/classification/dictionary_based/_tde.py:192]: Could you not use `**parameters` here to unnest and pass dictionary as keyword arguments to the constructor?

**Comment by mloning** [comment]
Hi @MatthewMiddlehurst, thanks for contributions, especially the notebook and unit tests, looks really good!

File: sktime/classification/dictionary_based/_tde.py (hunk 1 above)
Comment (MatthewMiddlehurst) [reply sktime/classification/dictionary_based/_tde.py]: Made a change making use of `*parameters` to clean it a bit. The parameter lists are fed to the GP  so cant be a dictionary (though now that i say that, i could have just extracted the values when they are added to the previous parameter list).

**Comment by mloning** [comment]
@MatthewMiddlehurst Is there an easy way to speed up the TDE unit tests while still checking actual values on the data sets?

**Comment by MatthewMiddlehurst** [comment]
Would you be able to change the parameters used? The way its setup with the GP and set number of parameters searched leads to it being comparatively poor on smaller datasets.

Accuracy doesn't really matter, if you used these it would be a lot faster while still running the important stuff:

n_parameter_samples=50
max_ensemble_size=10
randomly_selected_params=40

**Comment by mloning** [comment]
yes that would be fine, you can add it TDE to this dict here: https://github.com/alan-turing-institute/sktime/blob/7946e3588d0e526193be96ebd5ca250f07fb8890/sktime/tests/_config.py#L82

**Pull request merged**